sos/report/plugins/juju.py

```diff
@@ -53,6 +82,71 @@ def setup(self):
             # logs in the directory.
             self.add_copy_spec("/var/log/juju/*.log")
 
+        # Only include the Juju state report if this plugin option is set
+        if not self.get_option("juju-state"):
+            return
+
+        juju_user = self.get_option("juju-user")
+        try:
+            pwd.getpwnam(juju_user)
+        except KeyError:
+            self._log_warn(
+                f'User "{juju_user}" does not exist, '
+                "will not collect Juju information."
+            )
+            return
+
+        if self.get_option("controllers") and self.get_option("models"):
+            self._log_warn(
+                "Options: controllers, models are mutually exclusive. "
+                "Will not collect Juju information."
+            )
+            return
+
+        controllers_json = self.collect_cmd_output(
+            "juju controllers --format=json", runas=juju_user
+        )
+        if controllers_json["status"] == 0:
+            desired_controllers = set(
+                self.get_option("controllers").split(" ")
```

Comment:
We generally have plugins use `:` as a delimiter in plugin options that can take a list (as the use of commas gets murky with argparse and how plugin options are....parsed.

Reply:
The only problem with this is for the Juju model synatx it can be:

- The current controller's model -> `model_a`
- Or it can be a specific controller and model -> `controller_x:model_a`

For this reason I chose to split via empty space since models and controllers are not allowed to have spaces in their names according to Juju. Would this be a reasonable exception to the Sos plugin standard?

Reply:
If it's an expectation that users would pass something like `-k juju.controllers="foo:bar"`, then yeah it's totally reasonable to use something else. We should just make note of that, ideally in both the plugin option description (for `sos report -l` output) and the docstring for the plugin (as the docstring gets used in `sos help` output, e.g. `sos help report.plugins.juju` - see the `kernel` plugin docstring for a good example).

Reply:
I guess there is no reason why we couldn't use a colon delimiter for the controllers option. However, it is required for the models option (as I mentioned above) so I think it makes sense to be consistent within the scope of this plugin to keep the white-space delimiter rather than colon. What do you think?

Reply:
I added a docstring and the PluginOpts have `description` args. Is this enough to show up in the `sos report -l` output? I tested this locally and I could not get my Juju plugin options to show up in this output and I have no idea why.

[Reference](https://github.com/sosreport/sos/pull/3803/files#diff-3b9a49f615b1355c9acdf50238eb773562b9f1903d0c9ecbdbe5ae1eb3b6100aR70)

Reply:
This is what I get when I run your changes:

```
# bin/sos report -l | grep juju
 juju                 Juju orchestration tool
 juju.juju-state           off             Include Juju state in the report
 juju.juju-user            ubuntu          Juju client user.
 juju.controllers                          Collect Juju state for specified controllers. Uses a             whitespace delimited list.
 juju.models                               Collect Juju state for specified models. Uses a whitespace             delimited list.

```

And this is what I get when running sos help:

```
Plugin Options
These options may be toggled or changed using '-k juju.option_name=$value'

    Option Name         Default                       Description         
    juju-state          False/Off                     Include Juju state in the report
    juju-user           ubuntu                        Juju client user.
    controllers         None/Unset                    Collect Juju state for specified controllers. Uses a
                                                      whitespace delimited list.
    models              None/Unset                    Collect Juju state for specified models. Uses a whitespace
                                                      delimited list.

```

Reply:
Regarding why the options are not showing up in the output locally, perhaps it's because the machine where you are running doesn't have the /var/log/juju file present? Since that's the activation condition for this plugin:

`files = ('/var/log/juju',)`

As far as I understand it, if it's not present then the plugin should show as inactive and the options won't be printed.